src/encoded/static/components/matrix_chip_seq.js

```diff
@@ -0,0 +1,886 @@
+                )}
+            </div>
+        </ModalBody>
+    </Modal>);
+
+/**
+ * Container for ChIP-Seq Matrix page's content.
+ *
+ * @class ChIPSeqMatrixPresentation
+ * @extends {React.Component}
+ * @listens PubSub - SEARCH_PERFORMED_PUBSUB
+ */
+class ChIPSeqMatrixPresentation extends React.Component {
+    constructor(props) {
+        super(props);
+
+        this.subTabClicked = this.subTabClicked.bind(this);
+        this.performSearch = this.performSearch.bind(this);
+        this.handleOnScroll = this.handleOnScroll.bind(this);
+        this.handleScrollIndicator = this.handleScrollIndicator.bind(this);
+
+        const { context } = this.props;
+        const link = context['@id'];
+        const query = new QueryString(link);
+        const assayTitle = query.getKeyValues('assay_title')[0];
+        const organismName = query.getKeyValues('replicates.library.biosample.donor.organism.scientific_name')[0];
+        const selectedTab = (tabs.find(tab => tab.assayTitle === assayTitle && tab.organismName === organismName) || tabs[0]).title;
+
+        this.subTabs = [];
+        this.ChIPSeqMatrixData = [];
+
+        this.state = {
+            chIPSeqData: [],
+            selectedTab,
+            selectedSubTab: null,
+            scrolledRight: false,
+            showOrganismRequest: false,
+        };
+    }
+
+    componentDidMount() {
+        this.handleScrollIndicator(this.scrollElement);
+
+        // extract ChIP-Seq Matrix data and get relevant values out
+        const { context } = this.props;
+        const link = context['@id'];
+        const query = new QueryString(link);
+        const assayTitle = query.getKeyValues('assay_title')[0];
+        const organismName = query.getKeyValues('replicates.library.biosample.donor.organism.scientific_name')[0];
+        const showOrganismRequest = !(assayTitle && organismName);
+
+        // ALL ChIP-Seq Matrix data
+        this.ChIPSeqMatrixData = getChIPSeqData(context, assayTitle, organismName);
+
+        // sub tabs
+        this.subTabs = this.ChIPSeqMatrixData ? this.ChIPSeqMatrixData.subTabs : [];
+
+        // subtab may be in the url #, get it if it is there or default to first subtabs list value
+        const hash = window ? window.decodeURIComponent(window.location.hash).replace('#', '') : null;
+        const selectedSubTab = hash && this.subTabs.includes(hash) ?
+            hash :
+            this.subTabs.length > 0 ? this.ChIPSeqMatrixData.subTabs[0] : null;
+
+        // sub chIP Seq data to display
+        const chIPSeqData = this.ChIPSeqMatrixData ? this.ChIPSeqMatrixData.chIPSeqData[selectedSubTab] : {};
+        const isMatrixLarge = chIPSeqData && chIPSeqData.dataRow && chIPSeqData.dataRow.length > MIN_SIZE_OF_LARGE_MATRIX;
+        const matrixUpdate = {
+            chIPSeqData,
+            selectedSubTab,
+            showOrganismRequest, // determines if organism modal shows
+        };
+
+        if (isMatrixLarge) {
```

Comment:
Is this actually necessary to test if it's large?  If it's small it will just flash the spinner and move on, right?  Does this look stupid?

Reply:
It's silly and hacky but I had reasons. It is a rabbit hole. There is a spinner from fetched.js, which comes for free on every backend call. For small matrices, it suffices to use the spinner from fetched.js as it does not get clobbered by the matrix-painting. That is why I do not call showSpinner here on small size-matrices

On large matrices, I have to call _.defer() which pushes execution until call stack is empty. That is computationally expensive and should only be done when necessary. fetched.js usually kills the spinner in the time it takes to paint large matrices, so the matrices in this case has to be re-started by the painting-code and stopped afterwards. That means I have to check if the matrix is large and do things different compared to when small.

Reply:
good enough!